Merge CUDAFuture into ivalue::Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57052 Merge CUDAFuture into ivalue::Future**
* #57051 Make CUDAFuture handle any kind of device type
* #57050 Use only generic helpers in CUDAFuture
* #57049 Introduce generic MultiStreamGuard
* #57048 Simplify CUDAMultiStreamGuard
* #57047 Add recordDataPtrOnStream to DeviceGuardImplInterface
* #57046 Add getStreamFromPool to DeviceGuardImplInterface

This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

lw committed:
Merge CUDAFuture into ivalue::Future

This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -932,6 +932,34 @@ getClassConverter() {
 }
 
+// Needs to be in this .cpp file to access the full definition of PyObjectHolder
+std::vector<std::reference_wrapper<const at::DataPtr>> ivalue::Future::extractDataPtrs(
```
assuming no change in the implementation, except this is moved from `CUDAFuture`.

lw replied:
Yes I just copy-pasted.

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -311,7 +316,19 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
  public:
-  explicit Future(TypePtr type) : type_(type) {}
+  explicit Future(TypePtr type, std::vector<c10::Device> devices={})
+    : type_(std::move(type)),
```
this should be 4 spaces indent?

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -529,10 +551,44 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+      return;
+    }
+
+    // Start by performing all steps that can throw, before setting any field.
```
also assuming no change in the implementation below.

mrshenli reviewed (approved):
This is great! Looks like we don't have hard blockers for this stack. I am going to update and land #57085 after this one.

mrshenli reviewed (changes requested):
There are internal test failures.

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

mrshenli reviewed (approved):
stamp to unblock

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]

lw committed:
Update on "Merge CUDAFuture into ivalue::Future"


This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

[ghstack-poisoned]
